merino-suggest/src/providers/client_variant_filter.rs

```diff
@@ -0,0 +1,178 @@
+use crate::{CacheInputs, SuggestError, SuggestionProvider, SuggestionRequest, SuggestionResponse};
+use async_trait::async_trait;
+
+/// A provider that gives suggestions base
+pub struct ClientVariantFilterProvider {
+    matching_provider: Box<dyn SuggestionProvider>,
+    default_provider: Box<dyn SuggestionProvider>,
+    client_variant: String,
+}
+
+impl ClientVariantFilterProvider {
+    /// Create a boxed Client Variant Provider
+    pub fn new_boxed(
+        matching_provider: Box<dyn SuggestionProvider>,
+        default_provider: Box<dyn SuggestionProvider>,
+        client_variant: String,
+    ) -> Box<Self> {
+        Box::new(Self {
+            matching_provider,
+            default_provider,
+            client_variant,
+        })
+    }
+}
+
+#[async_trait]
+impl SuggestionProvider for ClientVariantFilterProvider {
+    fn name(&self) -> String {
+        format!(
+            "ClientVariant(matching:{}, default:{})",
+            self.matching_provider.name(),
+            self.default_provider.name()
+        )
+    }
+
+    async fn suggest(
+        &self,
+        request: SuggestionRequest,
+    ) -> Result<SuggestionResponse, SuggestError> {
+        let req = request.clone();
+        let provider = if req
+            .client_variants
+            .unwrap_or(vec![])
+            .contains(&self.client_variant)
+        {
+            &self.matching_provider
+        } else {
+            &self.default_provider
+        };
+        let results = provider
+            .suggest(request)
+            .await
+            .unwrap_or_else(|_| SuggestionResponse::new(vec![]));
+        Ok(results)
+    }
+
+    fn cache_inputs(&self, request: &SuggestionRequest, cache_inputs: &mut dyn CacheInputs) {
+        self.matching_provider.cache_inputs(request, cache_inputs);
+        self.default_provider.cache_inputs(request, cache_inputs);
+
+        let req = request.clone();
+
+        if req
+            .client_variants
+            .unwrap_or(vec![])
+            .contains(&self.client_variant)
+        {
+            cache_inputs
+                .add(format!("client_variant_match:{}=true", &self.client_variant).as_bytes());
+        } else {
+            cache_inputs
+                .add(format!("client_variant_match:{}=false", &self.client_variant).as_bytes());
+        };
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        CacheStatus, ClientVariantFilterProvider, SuggestError, Suggestion, SuggestionProvider,
+        SuggestionRequest, SuggestionResponse,
+    };
+    use async_trait::async_trait;
+    use fake::{Fake, Faker};
+
+    struct TestMatchSuggestionsProvider();
+
+    #[async_trait]
+    impl SuggestionProvider for TestMatchSuggestionsProvider {
```

Comment:
Instead of defining two bespoke providers, it *might* be easier to use the `Fixed` provider. It's a debugging provider where you specify a string in it's config, and it simply returns a suggestion with that string as the title. It's supposed to be for this kind of situation, but it hasn't been used much yet.